### PR TITLE
connectors: Add Unittests for opaconnector

### PR DESCRIPTION
**Pull request by rohithdv**
This PR addresses #124 . 

This PR does the following changes to opa connector:
(1) Minor redesign of opa connector to make it suitable for unit tests
(2) Preliminary unit tests for opa connector

To run the unit test, execute the following:
(1) cd to connectors/opa
(2)In the Makefile in this dir, set/unset NEED_TEST_COVERAGE_STATISTICS and NEED_TEST_COVERAGE_STATISTICS_IN_HTML as per requirement. 
(3) Execute make test 

Note: This PR is a revised version of PR https://github.com/IBM/the-mesh-for-data/pull/47 which was closed due to some errors due to DCO changes.

### CODE REVIEW

File: connectors/opa/lib/evaluate_opa.go [hunk 1]
```diff
@@ -57,17 +57,17 @@ func EvaluateExtendedPoliciesOnInput(inputMap map[string]interface{}, opaServerU
...
-	//fmt.Printf("body from policy http response: %s\n", data)
-	//fmt.Printf("status from policy http response: %d\n", res.StatusCode)
-	//res.Body.Close()
+	// }
+	// contentType := "application/text"
+	// res := performHTTPReq(standardClient, opaServerURL+"v1/policies/extendedEnforcement", httpMethod, string(content), contentType)
+	// data, _ := ioutil.ReadAll(res.Body)
+	// fmt.Printf("body from policy http response: %s\n", data)
+	// fmt.Printf("status from policy http response: %d\n", res.StatusCode)
+	// res.Body.Close()
```
Comment (shlomitk1) [inline connectors/opa/lib/evaluate_opa.go:70]: Can the commented code be removed?

**Review by shlomitk1 — APPROVED** [review]

File: connectors/opa/testutil/helper.go [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+func GetEnvironment() (int, string, string) {
+	timeOutInSecs := EnvValues["CONNECTION_TIMEOUT"]
+	timeOutSecs, _ := strconv.Atoi(timeOutInSecs)
+
+	catalogConnectorURL := EnvValues["CATALOG_CONNECTOR_URL"]
+	opaServerUrl := EnvValues["OPA_SERVER_URL"]
+	log.Printf("EnvVariables = %v\n", EnvValues)
+
+	return timeOutSecs, catalogConnectorURL, opaServerUrl
+}
```
Comment (roee88) [inline connectors/opa/testutil/helper.go:37]: Why not just use constants directly?

File: connectors/opa/testutil/helper.go [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+		enforcementActions := make([]*pb.EnforcementAction, 0)
+		usedPolicies := make([]*pb.Policy, 0)
+		operation = datasetContext.GetOperation()
+		fmt.Println("operation")
+		fmt.Println(operation)
+		var newUsedPolicy *pb.Policy
+		if purpose == "marketing" {
+			newEnforcementAction := ConstructRedactColumn("nameDest")
+			enforcementActions = append(enforcementActions, newEnforcementAction)
+			newUsedPolicy = &pb.Policy{Description: "reduct columns with name nameOrig and nameDest  in datasets with Finance"}
```
Comment (roee88) [inline connectors/opa/testutil/helper.go:94]: finance or marketing?

**Review by roee88 — COMMENTED** [review]
Added some comments. I'm still struggling with `testutil/helper.go`. It includes too much and actually having it hides what the test does (looking at the test file all you can tell is we send a request and expect some results, go to helper to see what)

File: connectors/opa/lib/opa_reader_test.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+// format into enforcement decisions format
+
+func TestMainOpaConnector(t *testing.T) {
+
+	timeOutSecs, catalogConnectorURL, opaServerUrl := tu.GetEnvironment()
+	applicationContext := tu.GetApplicationContext("marketing")
+
+	srv := NewOpaReader(opaServerUrl)
+	catalogReader := NewCatalogReader(catalogConnectorURL, timeOutSecs)
+	policiesDecisions, _ := srv.GetOPADecisions(applicationContext, catalogReader)
```
Comment (froesef) [inline connectors/opa/lib/opa_reader_test.go:27]: You should check that the error is nil:
```
policiesDecisions, err := srv.GetOPADecisions(applicationContext, catalogReader)
assert.NilError(t, err)
```

File: connectors/opa/main.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+		return nil, fmt.Errorf("Atoi conversion of timeOutinseconds failed: %v", err)
+	}
+
+	catalogReader := opabl.NewCatalogReader(catalogConnectorAddress, timeOut)
+	eval, err := s.opaReader.GetOPADecisions(in, catalogReader)
+	if err != nil {
+		log.Println("GetOPADecisions err:", err)
+		return nil, err
+	}
+	jsonOutput, _ := json.MarshalIndent(eval, "", "\t")
```
Comment (froesef) [inline connectors/opa/main.go:69]: Please handle the error here. Thus if there is an error you probably should not log the jsonOutput string.

File: connectors/opa/lib/evaluate_opa.go [hunk 6]
```diff
@@ -1,7 +1,7 @@
 // Copyright 2020 IBM Corp.
 // SPDX-License-Identifier: Apache-2.0
 
-package main
+package opaconnbl
```
[the remark this answers is not in the document]
Comment (rohithdv) [reply connectors/opa/lib/evaluate_opa.go:4]: sorry I didn't understand this comment. is it something to do with the naming convention of the package ?

File: connectors/opa/lib/opa_reader.go [hunk 7]
```diff
@@ -1,79 +1,31 @@
 // Copyright 2020 IBM Corp.
 // SPDX-License-Identifier: Apache-2.0
 
-package main
+package opaconnbl
```
[the remark this answers is not in the document]
Comment (rohithdv) [reply connectors/opa/lib/opa_reader.go:4]: sorry I didn't understand this comment. is it something to do with the naming convention of the package ?

File: connectors/opa/lib/catalog_reader.go [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+	if errJSON != nil {
+		log.Printf("error Marshalling Catalog External Connector Response: %v", errJSON)
+		return nil, fmt.Errorf("error Marshalling External Catalog Connector Response: %v", errJSON)
+	}
+	log.Print(string(responseBytes))
+	log.Println("***************************************************************")
+	metadataMap := make(map[string]interface{})
+	err = json.Unmarshal(responseBytes, &metadataMap)
+	if err != nil {
+		log.Printf("error in unmarshalling responseBytes: %v", err)
```
[the remark this answers is not in the document]
Comment (rohithdv) [reply connectors/opa/lib/catalog_reader.go]: fixed. Thanks.

File: connectors/opa/lib/catalog_reader.go [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+		log.Printf("error sending data to External Catalog Connector (datasetID = %s): %v", datasetID, err)
+		errStatus, _ := status.FromError(err)
+		log.Println("Message:", errStatus.Message())
+		log.Println("Code:", errStatus.Code())
+		if codes.InvalidArgument == errStatus.Code() {
+			log.Println("Invalid argument error : " + err.Error())
+		}
+		return nil, fmt.Errorf("error sending data to External Catalog Connector (datasetID = %s): %v", datasetID, err)
+	}
+	log.Println("***************************************************************")
```
[the remark this answers is not in the document]
Comment (rohithdv) [reply connectors/opa/lib/catalog_reader.go]: fixed. Thanks.

File: connectors/opa/lib/opa_reader_test.go (hunk 4 above)
Comment (rohithdv) [reply connectors/opa/lib/opa_reader_test.go:27]: Fixed by using the assert function from gotest.tools/assert. Thanks.

File: connectors/opa/main.go (hunk 5 above)
Comment (rohithdv) [reply connectors/opa/main.go]: Fixed. Thanks.

File: connectors/opa/testutil/helper.go (hunk 2 above)
Comment (rohithdv) [reply connectors/opa/testutil/helper.go:37]: having the EnvValues variable gives us flexibility to change these env variables in different testcases (if need be) as shown in example [here](https://github.com/rohithdv/the-mesh-for-data/blob/6bfa712c3ac2450a9f9eba2d4051384f9141b514/connectors/opa/lib/opa_reader_test.go#L43).

File: connectors/opa/testutil/helper.go (hunk 3 above)
Comment (rohithdv) [reply connectors/opa/testutil/helper.go:94]: we have been using 2 purposes by convention - fraud-detection and marketing in [other test cases](https://github.com/rohithdv/the-mesh-for-data/blob/6bfa712c3ac2450a9f9eba2d4051384f9141b514/pkg/policy-compiler/pc-bl/policy_compiler_test.go#L62). so we have used "marketing" here since in the opa connector test case context any purpose would do. these are just for testing purposes.

File: connectors/opa/lib/catalog_reader.go [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+			}
+			datasetsMetadata[datasetID] = metadataMap
+		}
+	}
+
+	return datasetsMetadata, nil
+}
+
+func (c *CatalogReader) GetDatasetMetadata(ctx *context.Context, client pb.DataCatalogServiceClient, datasetID string, appID string) (map[string]interface{}, error) {
+	objToSend := &pb.CatalogDatasetRequest{AppId: appID, DatasetId: datasetID}
```
[the remark this answers is not in the document]
Comment (rohithdv) [reply connectors/opa/lib/catalog_reader.go:61]: In the WKC context, all the connectors need the appID to retrieve the credentials for accessing WKC as shown [here](https://github.ibm.com/data-mesh-research/WKC-connector/blob/43100b73fb150c7acac7ac8dad8304f5fa0568d1/java/src/main/java/com/ibm/wkc/WKCCredentials.java#L69). 

In the open source context , I dont think we are using appID for retrieving credentials as shown [here](https://github.com/IBM/the-mesh-for-data/blob/0177f18d2772bb43c6c333209dc188e8d24570d1/connectors/vault/vault_connector.go#L32). Also we are NOT using appID to retrieve catalog information from the egeria catalog connector as shown [here](https://github.com/IBM/the-mesh-for-data/blob/0177f18d2772bb43c6c333209dc188e8d24570d1/connectors/egeria/src/main/java/com/ibm/egeria/EgeriaClient.java#L377). For Policy manager connector (OPA Connector) also we are NOT using appID as shown [here](https://github.com/IBM/the-mesh-for-data/blob/0177f18d2772bb43c6c333209dc188e8d24570d1/connectors/opa/opa_connector.go#L51). 

To summarise, we are using appID only in the context of WKC. We are not using appID when we are using open source components like opa connector, egeria connector and vault connector.

File: connectors/opa/Makefile [hunk 11]
```diff
@@ -7,9 +7,11 @@ include $(ROOT_DIR)/.mk/docker.mk
 
 OPA_PATH=${ABSTOOLBIN}/opa
 OPA_DEFAULT_POLICY_FILE=${ROOT_DIR}/third_party/opa/opa-policy.rego
+NEED_TEST_COVERAGE_STATISTICS=1
+NEED_TEST_COVERAGE_STATISTICS_IN_HTML=0
```
[the remark this answers is not in the document]
Comment (rohithdv) [reply connectors/opa/Makefile:11]: @hunchback Can you please comment on this? Thanks.

File: connectors/opa/lib/catalog_reader.go [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+	"log"
+	"time"
+
+	pb "github.com/ibm/the-mesh-for-data/pkg/connectors/protobuf"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// Reader from catalog, single instance for the connector, not dependent on the request
```
[the remark this answers is not in the document]
Comment (rohithdv) [reply connectors/opa/lib/catalog_reader.go]: We have separated the OPA connector functionality into 2 structs - CatalogReader (which has functions which reads info from the catalog) and OPAReader (which uses the catalog metadata and performs policy evaluation). 

CatalogReader is a struct which has information to read from catalog. This struct does not have any information related to the any specific application context. Any request specific info is passed as parameters to functions belonging to this struct. I have updated the comment of CatalogReader based on this explanation. Hope it is ok. 

I think both these structs are part of OPAConnector and they are just separated based on their distinct functionalities. In the earlier version of OpaConneector code, we had merged the functions related to catalog reader and opa reader which was a bit confusing. Now there are separated for better clarity.

[the remark this answers is not in the document]
Comment (rohithdv) [reply connectors/opa/lib/catalog_reader.go]: fixed. Please check if it is ok. Thanks.

**Comment by rohithdv** [comment]
> Added some comments. I'm still struggling with `testutil/helper.go`. It includes too much and actually having it hides what the test does (looking at the test file all you can tell is we send a request and expect some results, go to helper to see what)

basically testutil/helper.go has functions to (1) to start mock catalog connector (2) to start mock opa server (3) implements GetDatasetInfo on mock catalog connector (4) function to return mock OPA server response (5) have GetExpectedOpaDecisions() which returns pb.PoliciesDecisions object which has OPA decisions that are expected based on what are configured in the mocks.  (6) additional helper functions which help in implementing (1) - (5) above. 

At a high level, testutil/helper.go is designed to help us to configure different test cases based on different scenarios which may arise when working with opa connector.
 
Please let me know if this is ok. Thanks.

File: connectors/opa/lib/evaluate_opa.go (hunk 1 above)
Comment (rohithdv) [reply connectors/opa/lib/evaluate_opa.go]: fixed. Thanks.

File: connectors/opa/lib/catalog_reader.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (roee88) [reply connectors/opa/lib/catalog_reader.go]: What makes catalog reader specific to the opa connector. It looks like a generic client to a catalog connector

File: connectors/opa/lib/evaluate_opa.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (roee88) [reply connectors/opa/lib/evaluate_opa.go:4]: I assumed this is a leftover and you wanted it to match the directory name.

File: connectors/opa/testutil/helper.go (hunk 2 above)
Comment (roee88) [reply connectors/opa/testutil/helper.go:37]: Seems too complex but I assume it's a preparation for actually taking these values from environment variables.

File: connectors/opa/testutil/helper.go (hunk 3 above)
Comment (roee88) [reply connectors/opa/testutil/helper.go:94]: This is within a block of an if statement that compares to "marketing"

**Review by roee88 — APPROVED** [review]
Replied to your comments, let me know if you want to address them before merging.

File: connectors/opa/lib/evaluate_opa.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (rohithdv) [reply connectors/opa/lib/evaluate_opa.go]: yes right. Sorry I got confused with the older dir structure. I have changed the package name to lib now. Thanks.

File: connectors/opa/lib/opa_reader.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (rohithdv) [reply connectors/opa/lib/opa_reader.go]: I have changed the package name to lib now. Thanks.

File: connectors/opa/testutil/helper.go (hunk 2 above)
Comment (rohithdv) [reply connectors/opa/testutil/helper.go:37]: actually we wanted to not depend on env variables for running the test cases. hence we created this struct.

**Comment by rohithdv** [comment]
> Replied to your comments, let me know if you want to address them before merging.

Addressed most of your comments. pls let me know if I need to do anything about this comment. https://github.com/IBM/the-mesh-for-data/pull/125#discussion_r529367702. Thanks.

File: connectors/opa/testutil/helper.go (hunk 3 above)
Comment (rohithdv) [reply connectors/opa/testutil/helper.go]: Actually what the description means is “redact columns with name nameOrig and nameDest  in datasets which have been tagged with Finance” ..The purpose could be marketing or fraud-detection or any other purpose..I have edited the description to the above to avoid confusion. Thanks.

**Comment by rohithdv** [comment]
> > Replied to your comments, let me know if you want to address them before merging.
> 
> Addressed most of your comments. pls let me know if I need to do anything about this comment. [#125 (comment)](https://github.com/IBM/the-mesh-for-data/pull/125#discussion_r529367702). Thanks.

Actually what the description means is “redact columns with name nameOrig and nameDest  in datasets which have been tagged with Finance” ..The purpose could be marketing or fraud-detection or any other purpose..I have edited the description to the above to avoid confusion. Thanks.

File: connectors/opa/lib/catalog_reader.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (rohithdv) [reply connectors/opa/lib/catalog_reader.go]: ok thanks.
